File: packages/db/src/schema/modelLogs.ts
import {
  bigint,
  index,
  inet,
  integer,
  numeric,
  pgTable,
  serial,
  text,
  timestamp,
  unique,
  varchar,
} from "drizzle-orm/pg-core";

export const modelLogs = pgTable(
  "model_logs",
  {
    id: serial().primaryKey().notNull(),
    openUserId: varchar("open_user_id", { length: 32 }).default("").notNull(),
    providerCode: varchar("provider_code", { length: 50 })
      .default("")
      .notNull(),
    model: varchar({ length: 50 }).default("").notNull(),
    inputTokens: integer("input_tokens").default(0).notNull(),
    outputTokens: integer("output_tokens").default(0).notNull(),
    totalTokens: integer("total_tokens").default(0).notNull(),
    latencyMs: integer("latency_ms").default(0).notNull(),
    status: integer().default(0),
    errorMessage: text("error_message").default("").notNull(),
    // You can use { mode: "bigint" } if numbers are exceeding js number limitations
    cost: bigint({ mode: "number" }).default(0).notNull(),
    apiKey: varchar("api_key", { length: 255 }).default("").notNull(),
    ipAddress: inet("ip_address").default("0.0.0.0").notNull(),
    requestStartTime: timestamp("request_start_time", {
      withTimezone: true,
      mode: "string",
    }),
    responseCompleteTime: timestamp("response_complete_time", {
      withTimezone: true,
      mode: "string",
    }),
    tokensPerSecond: numeric("tokens_per_second", { precision: 10, scale: 2 }),
  },
  (table) => [
    index("idx_model_logs_model").using("btree", table.model.asc().nullsLast()),
    index("idx_model_logs_open_user_id").using(
      "btree",
      table.openUserId.asc().nullsLast()
    ),
    index("idx_model_logs_provider_code").using(
      "btree",
      table.providerCode.asc().nullsLast()
    ),
  ]
);

File: packages/go-shared/pkg/model/model_logs.go
package model

import (
	"time"
)

type ModelLog struct {
	ID                   int32      `gorm:"column:id;type:integer;primaryKey;autoIncrement:true" json:"id"`
	OpenUserID           string     `gorm:"column:open_user_id;type:character varying(32);not null;index:idx_model_logs_open_user_id,priority:1" json:"open_user_id"`
	ProviderCode         string     `gorm:"column:provider_code;type:character varying(50);not null;index:idx_model_logs_provider_code,priority:1" json:"provider_code"`
	Model                string     `gorm:"column:model;type:character varying(50);not null;index:idx_model_logs_model,priority:1" json:"model"`
	InputTokens          int32      `gorm:"column:input_tokens;type:integer;not null" json:"input_tokens"`
	OutputTokens         int32      `gorm:"column:output_tokens;type:integer;not null" json:"output_tokens"`
	TotalTokens          int32      `gorm:"column:total_tokens;type:integer;not null" json:"total_tokens"`
	LatencyMs            int32      `gorm:"column:latency_ms;type:integer;not null" json:"latency_ms"`
	Status               *int32     `gorm:"column:status;type:integer" json:"status"`
	ErrorMessage         string     `gorm:"column:error_message;type:text;not null" json:"error_message"`
	Cost                 int64      `gorm:"column:cost;type:bigint;not null" json:"cost"`
	APIKey               string     `gorm:"column:api_key;type:character varying(255);not null" json:"api_key"`
	IPAddress            *string    `gorm:"column:ip_address;type:inet;not null;default:0.0.0.0" json:"ip_address"`
	RequestStartTime     *time.Time `gorm:"column:request_start_time;type:timestamp with time zone" json:"request_start_time"`
	ResponseCompleteTime *time.Time `gorm:"column:response_complete_time;type:timestamp with time zone" json:"response_complete_time"`
	TokensPerSecond      *float64   `gorm:"column:tokens_per_second;type:numeric(10,2)" json:"tokens_per_second"`
}

func (ModelLog) TableName() string {
	return "model_logs"
}
